### PR TITLE
improve SHGetFileInfo

### PR DESCRIPTION
**Pull request by katahiromz**
This patch reduces failures of SHGetFileInfo function. CORE-7159

## Purpose

Improve SHGetFileInfo.
JIRA issue: [CORE-7159](https://jira.reactos.org/browse/CORE-7159)

## TODO

- [x] Reduce all failures of SHGetFileInfo in shell32_winetest:shlfileop.

### CODE REVIEW

**Comment by katahiromz** [comment]
Done. Review and commit me.

File: dll/win32/shell32/wine/shell32_main.c [hunk 1]
```diff
@@ -769,33 +794,25 @@ DWORD_PTR WINAPI SHGetFileInfoA(LPCSTR path,DWORD dwFileAttributes,
...
-                  psfi->szTypeName, sizeof(psfi->szTypeName), NULL, NULL);
-        }
+        WideCharToMultiByte(CP_ACP, 0, temppsfi.szDisplayName, -1,
+              psfi->szDisplayName, sizeof(psfi->szDisplayName), NULL, NULL);
+
+        WideCharToMultiByte(CP_ACP, 0, temppsfi.szTypeName, -1,
+              psfi->szTypeName, sizeof(psfi->szTypeName), NULL, NULL);
     }
+    else
+        ret = SHGetFileInfoW(pathW, dwFileAttributes, NULL, sizeof(temppsfi), flags);
```
Comment (SergeGautherie) [inline dll/win32/shell32/wine/shell32_main.c]: s/sizeof(temppsfi)/0/.

File: dll/win32/shell32/wine/shell32_main.c [hunk 2]
```diff
@@ -429,11 +430,11 @@ DWORD_PTR WINAPI SHGetFileInfoW(LPCWSTR path,DWORD dwFileAttributes,
         return FALSE;
 
     /* windows initializes these values regardless of the flags */
-    if (psfi != NULL)
+    if (psfi)
```
Comment (ThFabba) [inline dll/win32/shell32/wine/shell32_main.c]: Please avoid unnecessary style changes.

File: dll/win32/shell32/wine/shell32_main.c [hunk 3]
```diff
@@ -566,8 +591,8 @@ DWORD_PTR WINAPI SHGetFileInfoW(LPCWSTR path,DWORD dwFileAttributes,
 
     if (flags & SHGFI_LINKOVERLAY)
         uGilFlags |= GIL_FORSHORTCUT;
-    else if ((flags&SHGFI_ADDOVERLAYS) ||
-             (flags&(SHGFI_ICON|SHGFI_SMALLICON))==SHGFI_ICON)
+    else if ((flags & SHGFI_ADDOVERLAYS) ||
```
Comment (ThFabba) [inline dll/win32/shell32/wine/shell32_main.c]: More style changes

File: dll/win32/shell32/wine/shell32_main.c (hunk 1 above)
Comment (katahiromz) [reply dll/win32/shell32/wine/shell32_main.c]: Done.

File: dll/win32/shell32/wine/shell32_main.c (hunk 2 above)
Comment (katahiromz) [reply dll/win32/shell32/wine/shell32_main.c]: Done.

File: dll/win32/shell32/wine/shell32_main.c (hunk 3 above)
Comment (katahiromz) [reply dll/win32/shell32/wine/shell32_main.c]: Done.

**Comment by binarymaster** [comment]
Is this PR ready to squash & merge?

**Comment by yagoulas** [comment]
I think so. One thing that I still wonder about is if the fixes are supposed to be so simple, why didn't wine fix them already.

File: dll/win32/shell32/wine/shell32_main.c [hunk 4]
```diff
@@ -449,12 +450,24 @@ DWORD_PTR WINAPI SHGetFileInfoW(LPCWSTR path,DWORD dwFileAttributes,
...
     if (flags & SHGFI_EXETYPE)
     {
-        if (flags != SHGFI_EXETYPE)
-            return 0;
-        return shgfi_get_exe_type(szFullPath);
+        if (!(flags & SHGFI_SYSICONINDEX))
+        {
+            if (flags & SHGFI_USEFILEATTRIBUTES)
+            {
+                return 1;
```
Comment (binarymaster) [inline dll/win32/shell32/wine/shell32_main.c]: Some magic value here.

Is it SHGFI_SMALLICON (0x000000001)?

**Comment by katahiromz** [comment]
SHGetFileInfo function
https://msdn.microsoft.com/en-us/library/windows/desktop/bb762179.aspx

In this case, SHGFI_USEFILEATTRIBUTES and SHGFI_EXETYPE flags are used. MSDN says SHGFI_USEFILEATTRIBUTES and SHGFI_EXETYPE are mutually exclusive. So this case is abnormal. I don't know why it is TRUE.

**Comment by katahiromz** [comment]
SHGFI_SMALLICON is not for the return value. Can anyone understand?

**Comment by learn-more** [comment]
@katahiromz I do not understand your question.

**Comment by learn-more** [comment]
Probably because both are defined as 1?

`#define SHGFI_SMALLICON 1`

**Comment by katahiromz** [comment]
Yep.

**Comment by gedmurphy** [comment]
@yagoulas  @gigaherz @learn-more  unless you object, I'll merge them this weekend

**Comment by HBelusca** [comment]
@katahiromz : are there other things to add to this patch or is it ok to be merged?

**Comment by katahiromz** [comment]
@HBelusca: It's ready to be merged.

**Pull request merged**